File: epfl-hc/hc2014/drones2/DronesLib2.java
/////////////////////////////
// DO NOT SUBMIT THIS FILE //
/////////////////////////////
import java.util.*;
import java.io.*;
import java.awt.geom.Point2D;

public class DronesLib2 {
  /**
   * FUNCTIONS YOU MAY CALL
   *
   * write requested data to passed references, returns false once all
   * the points of a polygon are read
   */
  public static Point2D readOuter(int idx){
	  return uniRead(outer.get(idx));
  }
  
  public static Point2D readInner(int idx){
	  return uniRead(inner.get(idx));
  }
  
  /**
   * Functions you may call to visually debug your solution
   */
  public static void plotLine(Point2D a, Point2D b, int color){
    gpStream.printf("plot '-' w l lt %d\n%f %f\n%f %f\ne\n",
      color, a.getX(), a.getY(), b.getX(), b.getY());
  }

  public static void plotPoint(Point2D a, int color){
    gpStream.printf("plot '-' w p lt %d\n%f %f\ne\n",
      color, a.getX(), a.getY());
  }

  
  private static final int MIC = 0;
  private static final int MAC = 100;
  
  private static Vector<Queue<Point2D> > outer, inner;
  private static Point2D trashV = new Point2D.Double();
  private static double trashD;
  
  private static Point2D uniRead(Queue<Point2D> q){
    if (q.isEmpty()) return null;
    return q.poll();
  }
  
  /**
   * VISUALIZATION
   */
   
  private static Process gnuplot;
  private static PrintStream gpStream;

  public static void main(String args[]){
    int ntc,no,ni,nq;
	  double x,y,r;
	  Scanner sc = new Scanner(System.in);
	
	  // for each testcase
	  ntc = sc.nextInt();
	  for (int tc=0; tc<ntc; ++tc){
	    // init gnuplot
	    try{
  	    gnuplot = Runtime.getRuntime().exec("gnuplot --persist");
  	  }
  	  catch (IOException e){
  	    System.out.println(e.getMessage());
  	  }
	    gpStream = new PrintStream(gnuplot.getOutputStream(),true);
	    gpStream.printf("set terminal wxt size 900,900\n");
	    gpStream.printf("set multiplot\n");
	    gpStream.printf("unset key\n");
	    gpStream.printf("set xrange [%d:%d]\n",MIC,MAC);
	    gpStream.printf("set yrange [%d:%d]\n",MIC,MAC);
	
	    // read params
	    no = sc.nextInt();
	    ni = sc.nextInt();
	    nq = sc.nextInt();
	    outer = new Vector<Queue<Point2D> >(no);
	    inner = new Vector<Queue<Point2D> >(ni);
	
	    // read polygons
	    gpStream.printf("plot '-' w l lt 1\n");
	    for (int i=0; i<no; ++i){
	      outer.add(new LinkedList<Point2D>());
	      while (true){
		      x = sc.nextDouble(); y = sc.nextDouble();
		      if (x<0) break;
		      outer.get(i).add(new Point2D.Double(x,y));
		      gpStream.printf("%f %f\n",x,y);
	      }
	      gpStream.printf("%f %f\n\n",outer.get(i).peek().getX(), 
	          outer.get(i).peek().getY());
	    }
	    gpStream.printf( "e\n");
	    gpStream.printf("plot '-' w l lt 1\n");
	    for (int i=0; i<ni; ++i){
	      inner.add(new LinkedList<Point2D>());
	      while (true){
		      x = sc.nextDouble(); y = sc.nextDouble();
		      if (x<0) break;
		      inner.get(i).add(new Point2D.Double(x,y));
		      gpStream.printf("%f %f\n",x,y);
	      }
	      gpStream.printf("%f %f\n\n",inner.get(i).peek().getX(), 
	          inner.get(i).peek().getY());
	    }
	    gpStream.printf( "e\n");
	    
	    // call contestant init
	    Drones2.init(no,ni);
	
	    // query contestant
	    Vector<Double> contestantAnswers = new Vector<Double>();
	    gpStream.printf("plot '-' w p\n");
	    while (nq-- != 0){
		    x = sc.nextDouble(); y = sc.nextDouble();
		    gpStream.printf("%f %f\n",x,y);
		    contestantAnswers.add(Drones2.minRisk(new Point2D.Double(x,y)));
	    }
	    gpStream.printf( "e\n");
	    
	    // check contestant answers
	    for (int i=0; i<contestantAnswers.size(); ++i){
	      x = sc.nextDouble();
	      if (Math.abs(contestantAnswers.get(i)-x)>0.001){
	        System.out.printf("Wrong: Fails at query %d of test %d with %f " +
	            "instead of %f\n",i,tc,contestantAnswers.get(i),x);
	        return;
	      }
	    }
	  }
	  System.out.printf("Correct\n");
  }
}

File: epfl-hc/hc2014/drones2/dronesLib2.cpp
/////////////////////////////
// DO NOT SUBMIT THIS FILE //
/////////////////////////////

#include <iostream>
#include <cstdio>
#include <queue>
#include <vector>
#include <cmath>
#include "dronesLib2.hpp"
using namespace std;

// Minimal and maximal coordinate
#define MIC 0
#define MAC 100

vector<queue<Vec2d> > outer;
vector<queue<Vec2d> > inner;
Vec2d trashV;
double trashD;

/**
 * DATA INTERFACE
 */

bool uniRead(Vec2d &p, double &x, double &y, queue<Vec2d> &q){
	if (q.empty()) return false;
	p = q.front(); q.pop();
	x = p.x; y = p.y;
	return true;
}

bool readOuter(int idx, Vec2d &p){
	return uniRead(p, trashD, trashD, outer[idx]);
}
bool readOuter(int idx, double &x, double &y){
	return uniRead(trashV, x, y, outer[idx]);
}

bool readInner(int idx, Vec2d &p){
	return uniRead(p, trashD, trashD, inner[idx]);
}
bool readInner(int idx, double &x, double &y){
	return uniRead(trashV, x, y, inner[idx]);
}

/**
 * VISUALIZATION
 */
 
static FILE *gnuplot;

void plotLine(const Vec2d &a, const Vec2d &b, int color){
  fprintf(gnuplot, "plot '-' w l lt %d\n%lf %lf\n%lf %lf\ne\n",
    color, a.x, a.y, b.x, b.y);
}

void plotPoint(const Vec2d &a, int color){
  fprintf(gnuplot, "plot '-' w p lt %d\n%lf %lf\ne\n",
    color, a.x, a.y);
}


int main(){
	int ntc,no,ni,nq;
	double x,y;
	
	// for each testcase
	cin >> ntc;
	for (int tc=0; tc<ntc; ++tc){
	  // init gnuplot
	  gnuplot = popen("gnuplot --persist","w");
	  fputs("set terminal x11 size 900,900\n",gnuplot);
	  fputs("set multiplot\n",gnuplot);
	  fputs("unset key\n",gnuplot);
	  fprintf(gnuplot,"set xrange [%d:%d]\n",MIC,MAC);
	  fprintf(gnuplot,"set yrange [%d:%d]\n",MIC,MAC);
	
	  // read params
	  cin >> no >> ni >> nq;
	  outer.resize(no);
	  inner.resize(ni);
	
	  // read polygons
	  fprintf(gnuplot,"plot '-' w l lt 1\n");
	  for (int i=0; i<no; ++i){
	    while (true){
		    cin >> x >> y;
		    if (x<0) break;
		    outer[i].push(Vec2d(x,y));
		    fprintf(gnuplot,"%lf %lf\n",x,y);
	    }
	    fprintf(gnuplot,"%lf %lf\n\n",outer[i].front().x, outer[i].front().y);
	  }
	  fprintf(gnuplot, "e\n");
	  fprintf(gnuplot,"plot '-' w l lt 1\n");
	  for (int i=0; i<ni; ++i){
	    while (true){
		    cin >> x >> y;
		    if (x<0) break;
		    inner[i].push(Vec2d(x,y));
		    fprintf(gnuplot,"%lf %lf\n",x,y);
	    }
	    fprintf(gnuplot,"%lf %lf\n\n",inner[i].front().x, inner[i].front().y);
	  }
	  fprintf(gnuplot, "e\n");
	
	  // call contestant init
	  init(no,ni);
	
	  // query contestant
	  vector<double> contestantAnswers;
	  fprintf(gnuplot,"plot '-' w p\n");
	  while (nq--){
		  cin >> x >> y;
		  fprintf(gnuplot,"%lf %lf\n",x,y);
		  contestantAnswers.push_back(minRisk(Vec2d(x,y)));
	  }
	  fprintf(gnuplot, "e\n");
	  
	  // check contestant answers
	  for (int i=0; i<contestantAnswers.size(); ++i){
	    cin >> x;
	    if (fabs(contestantAnswers[i]-x)>0.001){
	      printf("Wrong: Fails at query %d of test %d with %lf instead of %lf\n",
	          i,tc,contestantAnswers[i],x);
	      return 0;
	    }
	  }
	}
	printf("Correct\n");
	return 0;
}
